app/models/sms/decoder.rb

```diff
@@ -97,6 +101,32 @@ def find_form
       @questionings = @form.smsable_questionings
     end
 
+    def find_user
+      @user = @msg.user
+      raise_decoding_error("user_not_found") unless @user && @user.active?
+    end
+
+    def set_auth_code
+      @auth_code = @tokens.shift.downcase if @tokens.first =~ /[a-z0-9]{4}/i
+    end
+
+    def authenticate_user
+      if @form.authenticate_sms
+        log_authentication_failure unless @auth_code.present? && @user.sms_auth_code == @auth_code
+      end
+    end
+
+    def log_authentication_failure
+      @msg.auth_failed = true
+      @msg.save(validate: false)
+      raise_decoding_error("user_not_found")
+    end
+
+    def check_for_brute_force
+      messages =  Sms::Message.where(user: @user).since(1.minute.ago).where(auth_failed: true)
+      raise_decoding_error("account_locked") if messages.count >= 3
```

Comment:
Can we move `3` and `1.minute` into constants? Also I realized that my spec was a little weird. I like the way you've implemented it. I think that the minimum interval for brute force tries would equal the lookback window divided by the threshold. Right now that is 1 minute / 3 == 20 seconds. Probably fine. But good to have these as constants: `BRUTE_FORCE_CHECK_WINDOW` and `BRUTE_FORCE_LOCK_THRESHOLD` or something. Perhaps also put the formula I'm suggesting in a comment above it. (If you think I'm correct!)